bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java

```diff
@@ -0,0 +1,296 @@
+
+    private SerialPortIdentifier portId;
+    private SDS011Handler thingHandler;
+    private @NonNullByDefault({}) SerialPort serialPort;
+
+    private @NonNullByDefault({}) OutputStream outputStream;
+    private @NonNullByDefault({}) InputStream inputStream;
+
+    public SDS011Communicator(SDS011Handler thingHandler, SerialPortIdentifier portId) {
+        this.thingHandler = thingHandler;
+        this.portId = portId;
+    }
+
+    /**
+     * Initialize the communication with the device, i.e. open the serial port etc.
+     *
+     * @param mode the {@link WorkMode} if we want to use polling or reporting
+     * @param interval the time between polling or reportings
+     * @return {@code true} if we can communicate with the device
+     * @throws PortInUseException
+     * @throws TooManyListenersException
+     * @throws IOException
+     * @throws UnsupportedCommOperationException
+     */
+    public boolean initialize(WorkMode mode, Duration interval)
+            throws PortInUseException, TooManyListenersException, IOException, UnsupportedCommOperationException {
+        boolean initSuccessful = true;
+
+        SerialPort localSerialPort = portId.open(thingHandler.getThing().getUID().toString(), 2000);
+        localSerialPort.setSerialPortParams(9600, 8, 1, 0);
+
+        outputStream = localSerialPort.getOutputStream();
+        inputStream = localSerialPort.getInputStream();
+
+        if (inputStream == null || outputStream == null) {
+            throw new IOException("Could not create input or outputstream for the port");
+        }
+
+        // wake up the device
+        initSuccessful &= sendSleep(false);
+        initSuccessful &= getFirmware();
+
+        if (mode == WorkMode.POLLING) {
+            initSuccessful &= setMode(WorkMode.POLLING);
+            initSuccessful &= setWorkingPeriod((byte) 0);
+        } else {
+            // reporting
+            initSuccessful &= setWorkingPeriod((byte) interval.toMinutes());
+            initSuccessful &= setMode(WorkMode.REPORTING);
+        }
+
+        // enable listeners only after we have configured the sensor above because for configuring we send and read data
+        // sequentially
+        localSerialPort.notifyOnDataAvailable(true);
+        localSerialPort.addEventListener(this);
+        this.serialPort = localSerialPort;
+
+        return initSuccessful;
+    }
+
+    private @Nullable SensorReply sendCommand(CommandMessage message) throws IOException {
+        byte[] commandData = message.getBytes();
+        if (logger.isDebugEnabled()) {
+            logger.debug("Will send command: {} ({})", Helper.toHexString(commandData), Arrays.toString(commandData));
+        }
+        outputStream.write(commandData, 0, commandData.length);
+        outputStream.flush();
+        try {
+            // Give the sensor some time to handle the command
+            Thread.sleep(500);
```

Comment:
Unfortunately this doesn't work. It is probably a "feature" of the sensor. So I have to wait here before reading the reply, otherwise the communication with the sensor will be broken.

Reply:
I'm still confused, how does reading data from a serial port break communication with the sensor. Are you sure this isn't a bug in the gnu.io serial library?